### PR TITLE
Add External Shaderpack support and add cloud/block outline shaders

### PR DESCRIPTION
**Pull request by OverlordsIII**
This pr adds external shaderpack support and adds a couple of new shaders for clouds and block outlines.
 * If this pr is accepted, when you start up minecraft, Iris will crash asking you to change your shaderpack in `config/iris.properties`. Simply change the end value of `shaderpack` to be the unzipped folder of the shaderpack. Make sure that the unzipped shaderpack has the `shader` directory directly inside of it.
* It also adds shaders for clouds and block outlines (basic)
* adds parsing of `item.properties` files and `block.properties` files in shaderpacks. Will be used later for uniforms and attributes. Can return the item and block map by calling `Iris.getShaderParser().getItemProperties()` or `Iris.getShaderParser().getBlockProperties()`

I probably messed something up, so please tell me what is wrong so I can fix it.
Thanks for letting me pr on this project.

### CODE REVIEW

**Review by IMS212 — CHANGES_REQUESTED** [review]
you should probably catch if the shaderpack is missing and provide a more user friendly message, and allow the user to change the file to load instead of hardcoding it to gbuffers_basic

**Comment by IMS212** [comment]
Otherwise, outside the already known water issue, the Trippy shaderpack seems to work fine.

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Thanks for the PR! I've left some comments from a first pass over the code.

In general I prefer for PRs to be split up into one PR per logical feature. I think the external shaderpack loading is a good improvement, but I have a different approach that I'd like to take for the clouds / outline shaders. I'll test it out locally and then push it, then the PR can be modified accordingly.

**Comment by ColdIce1605** [comment]
I've made some changes to this pr to reflect the changes requested but I don't want to summit a pr for it because of obvious reasons(ripping off it).
its also updated so there is no merge conflicts.

**Comment by OverlordsIII** [comment]
So this pr is pretty much done, I just have to figure out how to fix the block outline shaders from not working properly, but besides that I've implemented all the changes you requested.

**Comment by ColdIce1605** [comment]
hahahaha lmao

**Comment by OverlordsIII** [comment]
> hahahaha lmao

I definitely have no idea what you are talking about lol

**Comment by ColdIce1605** [comment]
I've thought you left this pr, sorry. So I started to solve the problems with this pr.

**Comment by OverlordsIII** [comment]
> I've thought you left this pr, sorry. So I started to solve the problems with this pr.
oh uhh
i guess that's what you get when you never commit anything till its done lol

**Comment by ColdIce1605** [comment]
yeah

**Comment by ColdIce1605** [comment]
once this pr is accepted 
I'll do a follow up that adds Logging for errors for shader devs and just the mod itself

**Comment by OverlordsIII** [comment]
Okay that makes sense
I feel a little wierd using println in this mod too
I think i'll also probably remove the outline shaders from this pr and focus on external loading

**Comment by ColdIce1605** [comment]
Is it suppose to compile?

**Comment by OverlordsIII** [comment]
Umm yea it’s supposed to compile unless I did something really bad.... what is it saying?

**Comment by ColdIce1605** [comment]
`C:\Users\James_R\Documents\Projects\Iris-master\src\main\java\net\coderbot\iris\mixin\MixinWorldRenderer.java:94: error: cannot find symbol
                Iris.getPipeline().beginBasic();
                                  ^
  symbol:   method beginBasic()
  location: class ShaderPipeline
C:\Users\James_R\Documents\Projects\Iris-master\src\main\java\net\coderbot\iris\mixin\MixinWorldRenderer.java:98: error: cannot find symbol
                Iris.getPipeline().endBasic();
                                  ^
  symbol:   method endBasic()
  location: class ShaderPipeline
Note: C:\Users\James_R\Documents\Projects\Iris-master\src\main\java\net\coderbot\iris\pipeline\ShaderPipeline.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
2 errors
`

**Comment by OverlordsIII** [comment]
Oops uh it should work now... sorry about that

**Comment by ColdIce1605** [comment]
https://imgur.com/syN7X2A
first time its worked for me.

**Comment by ColdIce1605** [comment]
I know this is a little outta scope(PropertiesParser) but optifine shaders also have shader.properties this should also be under the PropertiesParser class.

**Comment by OverlordsIII** [comment]
> I know this is a little outta scope(PropertiesParser) but optifine shaders also have shader.properties this should also be under the PropertiesParser class.

Yeah i am actually going to add it today or tomorrow, just trying to see what all the options are

File: src/main/java/net/coderbot/iris/shaderpack/PropertiesParser.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * a map that contains the identifier of an item to the integer value parsed in item.properties
+     */
+    private Map<Identifier, Integer> itemPropertiesMap = Maps.newHashMap();
+    private Map<Identifier, RenderLayer> blockRenderLayerMap = Maps.newHashMap();
+
+    public PropertiesParser(ShaderPack pack){
+        this.shaderPath = pack.getPath();
+        parseBlockProperties();
+        parseItemProperties();
```
Comment (ColdIce1605) [inline src/main/java/net/coderbot/iris/shaderpack/PropertiesParser.java:32]: add shaderProperties per optifine's spec.

**Review by ColdIce1605 — CHANGES_REQUESTED** [review]
some things need to be added in PropertiesParser.java

**Comment by coderbot16** [comment]
I'd rather not expand the scope of this PR too much, please. Just the block and item properties parsing is fine for now.

**Comment by OverlordsIII** [comment]
> I'd rather not expand the scope of this PR too much, please. Just the block and item properties parsing is fine for now.

ok yea, sorry about that. I'll just table it for now

**Comment by ColdIce1605** [comment]
anything keeping this pr from being merged?

**Review by WalkerKnapp — CHANGES_REQUESTED** [review]
I just suggested a few formatting changes, but how `IrisConfig` is setup is confusing.

File: src/main/java/net/coderbot/iris/pipeline/ShaderPipeline.java [hunk 2]
```diff
@@ -1,17 +1,17 @@
...
+import net.minecraft.client.render.RenderLayer;
 import org.jetbrains.annotations.Nullable;
 import org.lwjgl.opengl.GL20;
 
-import net.minecraft.client.gl.GlProgramManager;
-import net.minecraft.client.render.RenderLayer;
+import java.io.IOException;
+import java.util.Objects;
+
+;
```
[the remark this answers is not in the document]
Comment (OverlordsIII) [reply src/main/java/net/coderbot/iris/pipeline/ShaderPipeline.java:14]: Its my ide, it keeps rearranging the imports whenever I open up a class... It's really annoying. I'll try to disable it

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * A class dedicated to storing the config values of shaderpacks. Right now it only stores the path to the current shaderpack
+ */
+public class IrisConfig {
+    private Map<String, String> stringEntries = Maps.newHashMap();
+    private Map<String, Integer> intEntries = Maps.newHashMap();
+    private Map<String, Boolean> boolEntries = Maps.newHashMap();
```
[the remark this answers is not in the document]
Comment (OverlordsIII) [reply src/main/java/net/coderbot/iris/config/IrisConfig.java:19]: Its just set up that way so you wouldn't need to cast to make sure your object is actually that object. For other objects there will then be another map or properties object for just objects that you could use for things like shaderpath

File: src/main/java/net/coderbot/iris/pipeline/ShaderPipeline.java [hunk 4]
```diff
@@ -199,4 +198,15 @@ public void beginWorldRender() {
...
+
+	public void endBasic() {
+		GlProgramManager.useProgram(0);
+	}
+
+	public void beginBasic() {
+		if (basic == null){
+			return;
+		}
+		basic.use();
```
[the remark this answers is not in the document]
Comment (ColdIce1605) [reply src/main/java/net/coderbot/iris/pipeline/ShaderPipeline.java:210]: why not also do that for texturedLit

[the remark this answers is not in the document]
Comment (WalkerKnapp) [reply src/main/java/net/coderbot/iris/pipeline/ShaderPipeline.java:210]: Hm, it looks like all of the methods in `ShaderPipeline` are structured like that. Not sure why, but it's probably best to not change it.

File: src/main/java/net/coderbot/iris/Iris.java [hunk 5]
```diff
@@ -61,16 +72,61 @@ private static Program createShaders() {
 
 		return builder.build();
 	}
-
 	@Override
 	public void onInitializeClient() {
+		ShaderProperties properties = new ShaderProperties().setDefaultPack("Vaporwave-Shaderpack-master");
```
[the remark this answers is not in the document]
Comment (OverlordsIII) [reply src/main/java/net/coderbot/iris/Iris.java]: should be fixed by now

**Comment by OverlordsIII** [comment]
uh just got the shaders for the outline of the block to work. sorry that i couldn't get it in earlier

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        if (!propertiesPath.toFile().exists()){
+            return;
+        }
+        Properties properties = new Properties();
+        properties.load(Files.newInputStream(propertiesPath));
+        properties.forEach((key, value) -> {
+           if (key.toString().equals("shaderpack")){
+                this.shaderpath = Paths.get(value.toString());
+            }
+        });
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/config/IrisConfig.java:103]: No need to iterate over every property, just query the one that you want.
```suggestion
        this.shaderpath = Paths.get(properties.getProperty("shaderpack", "internal"));
```

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+     * Serializes the config into a file. Should be called after refreshing the values of entries
+     * @throws IOException file exceptions
+     */
+
+    public void serialize() throws IOException {
+        Properties properties = new Properties();
+        if (!shaderpath.toFile().exists()){
+            if (!shaderpath.endsWith("internal")) {
+                System.out.println(String.format("The specified shaderpack \"%s\" was not found! Change the value in iris.properties in your config directory! The system path should be \"%s\"", shaderpath.getFileName(), shaderpath));
+                System.out.println("falling back to internal shaders...");
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/config/IrisConfig.java:116]: Use System.err?

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+    /**
+     * Serializes the config into a file. Should be called after refreshing the values of entries
+     * @throws IOException file exceptions
+     */
+
+    public void serialize() throws IOException {
+        Properties properties = new Properties();
+        if (!shaderpath.toFile().exists()){
+            if (!shaderpath.endsWith("internal")) {
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/config/IrisConfig.java:114]: Why not equals?

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    /**
+     * Serializes the config into a file. Should be called after refreshing the values of entries
+     * @throws IOException file exceptions
+     */
+
+    public void serialize() throws IOException {
+        Properties properties = new Properties();
+        if (!shaderpath.toFile().exists()){
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/config/IrisConfig.java:113]: Avoid toFile whenever possible! Use the correct method in [`Files`](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html) instead. As far as I know, attempting to convert a valid Path to a File when the Path is ie. in a custom filesystem (like inside a jar file) will crash.

In this case, use `File#exists`.

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+            if (!shaderpath.endsWith("internal")) {
+                System.out.println(String.format("The specified shaderpack \"%s\" was not found! Change the value in iris.properties in your config directory! The system path should be \"%s\"", shaderpath.getFileName(), shaderpath));
+                System.out.println("falling back to internal shaders...");
+            }
+            shaderpath = FabricLoader.getInstance().getModContainer("iris")
+                    .orElseThrow(() -> new RuntimeException("Iris doesnt exists anymore I guess")).getRootPath();
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/config/IrisConfig.java:119]:
```suggestion
                    .orElseThrow(() -> new RuntimeException("Failed to get the mod container for Iris!")).getRootPath();
```

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            isInternal = true;
+        }
+        properties.setProperty("shaderpack", shaderpath.toString());
+        properties.store(Files.newOutputStream(propertiesPath), "This file is used to parse Iris Shader config options");
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/config/IrisConfig.java:123]:
```suggestion
        properties.store(Files.newOutputStream(propertiesPath), "This file stores configuration options for Iris, such as the currently active shaderpack");
```

File: src/main/java/net/coderbot/iris/config/IrisConfig.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * This sets the default shaderpack name for the shader properties to set.
+     * This value will only be used if the iris.properties file is not found
+     * @param shaderPackName the file name of the pack
+     * @return the shaderproperties with the set value
+     */
+    public IrisConfig setDefaultPack(String shaderPackName){
+        this.shaderpath = FabricLoader.getInstance().getGameDir().resolve("shaderpacks").resolve(shaderPackName);
+        return this;
+    }
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/config/IrisConfig.java:40]: I think that the internal shaders should always be the default, this method probably doesn't need to be exposed publicly.

**Review by coderbot16 — CHANGES_REQUESTED** [review]

**Review by WalkerKnapp — APPROVED** [review]
Everything I had comments on looks good 👍

**Review by coderbot16 — CHANGES_REQUESTED** [review]
A few small changes. I also see some assorted minor stylistic issues throughout the PR, but I'll just fix them myself. Once these remaining things are addressed the PR will be good to go!

**Review by coderbot16 — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**